### PR TITLE
Utility Functions Library

### PR DESCRIPTION
**Pull request by tsa96**
I keep meaning to put this together, finally got on and did it.

Does a lot of the work towards https://github.com/momentum-mod/website/issues/747, but don't want to spend more than a day or this so leaving the rest for now.

Main thing is `expandToIncludes` which we use *constantly* on GET endpoints, been wanting to consolidate a lot of similar logic into one place for a while now.

### CODE REVIEW

**Comment by tsa96** [comment]
One small build issue still, will fix tomorrow. Can still be reviewed in meantime

**Review by Gocnak — CHANGES_REQUESTED** [review]

File: apps/backend/src/app/modules/maps/map-review.service.ts [hunk 1]
```diff
@@ -33,13 +33,12 @@ export class MapReviewService {
...
     const hasRoleFiltering = query.official !== undefined;
     if (hasRoleFiltering) {
-      include ??= {};
       include['reviewer'] = true;
     }
 
     const dbResponse = await this.db.mapReview.findMany({
       where: { mapID },
-      include
+      include: trueIfEmpty(include)
```
[the remark this answers is not in the document]
Comment (tsa96) [reply apps/backend/src/app/modules/maps/map-review.service.ts]: Hmm, are you reviewing an out of date branch somehow? You're right, this doesn't make sense, a Prisma include is typically like
```ts
include: {
  model1: true;
  model2: { include: { model3: true } }
}
```
Using `true` in on the outermost `include` makes no sense and should be a type error.

Occasionally it makes sense to use `trueIfEmpty` for something like
```ts
include: {
  mmap: someExtraCondition ? { info:  true; } : true
}
```
where if `someExtraCondition` is false we just use `true` to only include `mmap` with no nested included.

But we were only actually using it in one place and I realised it's probably easier to just use `isEmpty(include) ? true : { include: include }` in such cases - an extra function feels like overkill and you have need to be quite familiar with that Prisma behaviour, so I've just been using the ternary expression on the last branch. Though, I didn't actually remove the function, which I may as well do given that it's being unused, will do that now.

`undefinedIfEmpty` is genuinely useful, mind, as passing Prisma an `include: {}` is an error (it's really annoying) so we use it a lot (8 usages currently).

[the remark this answers is not in the document]
Comment (tsa96) [reply apps/backend/src/app/modules/maps/map-review.service.ts]: Oh I'm sorry, I messed up my commit order a bit trying to split the `isEmpty` and `expandToIncludes` stuff. The `expandToIncludes` commit fixes this.  I can do some Git witchcraft to consolidate these if you want, though if you check in "File Changed" you'll see this and your above comment are fixed.

File: apps/backend/src/app/modules/users/users.service.ts [hunk 2]
```diff
@@ -667,7 +663,9 @@ export class UsersService {
     if (search) where.mmap = { name: { contains: search } };
 
     const include: Prisma.MapFavoriteInclude = {
-      user: true
+      user: true,
+      mmap: trueIfEmpty(
+      )
```
[the remark this answers is not in the document]
Comment (tsa96) [reply apps/backend/src/app/modules/users/users.service.ts]: Woops, see below comment

File: apps/backend/src/app/modules/maps/map-review.service.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Gocnak) [reply apps/backend/src/app/modules/maps/map-review.service.ts]: Ahhh I see. Yeah in the future try to make commit by commit stuff clean, I usually do that to review things.

**Review by Gocnak — APPROVED** [review]

**Pull request merged**